### ISSUE TITLE
Support for opfs in browsers

### ISSUE THREAD

**Issue by kesavkolla** [electric-sql/pglite#9]
For browsers it would be better to add support for OPFS so that we are not bounded by the memory limits and also faster than indexdb.  Infact there is no serialization of data to be done of we use the filesystem that is available. OPFS is supported in all major browsers. emscripten also supports wasmfs which can be used with OPFS

**Comment by samwillis** [comment]
Hey, yes OPFS is absolutely on the roadmap. IndexedDB was the quickest route to this early Alpha, but it's not the target.

**Comment by kesavkolla** [comment]
Yeah indexdb is ok but not good for performance. OPFS gives that performance boost. It also has sync access so it makes so much sense for Postgresql to use it. We don't need that async layer to deal with.

**Comment by samwillis** [comment]
Hey, yep exactly. It's the key target for persistence.

There are a few technicalities, although OPFS offers sync file handles, opening a file is still async. There a a few tricks we can take to get around this, using asyncify (which we currently use but is a resource overhead), or keeping a pool of open files.

lots to explore.

**Comment by AntonOfTheWoods** [comment]
Probably the work done by Roy Hashimoto over on https://github.com/rhashimoto/wa-sqlite/ could get someone who knows what they are doing most of the way there!

While it obviously has some drawbacks, given the now wide compatibility... I think spinning up a web worker and treating the worker process like a browser-internal db server is pretty elegant. This approach seems to work well with `wa-sqlite`, and (with some admittedly hacky workarounds) means you can get multiple tabs talking to a single web worker. It also means you can easily use the sync API and spinning up and waiting for the pool of handles just gets done once (for most circumstances probably).

**Comment by AntonOfTheWoods** [comment]
Unless mistaken, `wasmfs` is maybe not the best target for `opfs` sync, at least according to https://sqlite.org/wasm/doc/trunk/persistence.md#vfs-opfs-sahpool and the `wa-sqlite` discussions. Given it looks like pglite is going to necessarily be single process in accessing the files anyway and there is already a worker option, maybe a better target (in addition to the existing works-everywhere idb option) would be something mirroring the `sqlite` `vfs-opfs-sahpool` or `wa-sqlite` `AccessHandlePool`?

**Comment by samwillis** [comment]
Hey @AntonOfTheWoods, I expect we are going to explore both routes in time. WASMFS is probably the shorter route to a first version, but may have drawbacks with CORS header requirements.
An access handle pool would work, but would probably be a new Emscripten file system, Postgres doesn't have quite the same VFS concept as SQLite so we can't do it quite the same way.
Postgres also uses a lot more file than SQLite, so its not having a handful of access handles, it's potentially hundreds.

**Comment by Neo-Ciber94** [comment]
Would be great to just bring your own fs instead of using predefined ones, so we could bring our own.

```ts
abstract class FileSystem {
    abstract readFile(path: string): Promise<string>;
    abstract writeFile(path: string, contents: string): Promise<void>;
    /// other fs methods
}
```

```ts
import { PGLite } from "@electric-sql/pglite";
import { MemoryFS, NodeFS, IndexedDbFS  } from "@electric-sql/pglite/fs";

const pg = new PGLite(new MemoryFS());
```

**Comment by samwillis** [comment]
@Neo-Ciber94 absolutely, that's something we are considering. All the FS are built on top of the Emscripten file systems at the moment, and there aren't really any unofficial ones out there. We would need to find the right level of abstraction to enable custom VFSs but it's certainly something that make a lot of sense to do!

**Comment by uma-uc** [comment]
Any updates on this ?

**Comment by nikelborm** [comment]
Seems like it's done, because `packages/pglite/examples` has [opfs.html](https://github.com/electric-sql/pglite/blob/main/packages/pglite/examples/opfs.html) and [opfs-worker.js](https://github.com/electric-sql/pglite/blob/main/packages/pglite/examples/opfs-worker.js)

**Comment by AntonOfTheWoods** [comment]
@nikelborm is it using `wasmfs` though?

**Comment by nikelborm** [comment]
> [@nikelborm](https://github.com/nikelborm) is it using `wasmfs` though?

No idea, I just saw those 2 files in repo and mentioned it, thinking it may be related. I haven't used OPFS.

**Comment by samwillis** [comment]
Hey @nikelborm, the OPFS VFS is documented here: https://pglite.dev/docs/filesystems#opfs-ahp-fs

It's not an ideal implementation yet, and does not use wasmfs (something that comes with Emscripten). We plan to write a new OPFS VFS at some point, but there are a few problems we need to solve first to make it optimal for use in all browsers.

We may not use the wasmfs from Emscripten as it adds some overhead and complexity to the build presses and deployment. We are more likely to write a new one from scratch that tailored to PGlite. It's on the to do list.
